sql/core/src/main/scala/org/apache/spark/sql/Observation.scala

```diff
@@ -0,0 +1,170 @@
+ *   val observation = Observation("my_metrics")
+ *   val observed_ds = ds.observe(observation, count(lit(1)).as("rows"), max($"id").as("maxid"))
+ *   observed_ds.write.parquet("ds.parquet")
+ *   val metrics = observation.get
+ * }}}
+ *
+ * This collects the metrics while the first action is executed on the obseerved dataset. Subsequent
+ * actions do not modify the metrics returned by [[org.apache.spark.sql.Observation.get]]. Retrieval
+ * of the metric via [[org.apache.spark.sql.Observation.get]] blocks until the first action has
+ * finished and metrics become available. You can add a timeout to that blocking via
+ * [[org.apache.spark.sql.Observation.waitCompleted]]:
+ *
+ * {{{
+ *   if (observation.waitCompleted(100, TimeUnit.MILLISECONDS)) {
+ *     observation.get
+ *   }
+ * }}}
+ *
+ * This class does not support streaming datasets.
+ *
+ * @param name name of the metric
+ * @since 3.2.0
+ */
+class Observation(name: String) {
+
+  private val listener: ObservationListener = ObservationListener(this)
+
+  private var sparkSession: Option[SparkSession] = None
+
+  @volatile private var row: Option[Row] = None
+
+  /**
+   * Attaches this observation to the given [[Dataset]] to observe aggregation expressions.
+   *
+   * @param ds dataset
+   * @param expr first aggregation expression
+   * @param exprs more aggregation expressions
+   * @tparam T dataset type
+   * @return observed dataset
+   * @throws IllegalArgumentException If this is a streaming Dataset (ds.isStreaming == true)
+   */
+  def on[T](ds: Dataset[T], expr: Column, exprs: Column*): Dataset[T] = {
+    if (ds.isStreaming) {
+      throw new IllegalArgumentException("Observation does not support streaming Datasets")
+    }
+    register(ds.sparkSession)
+    ds.observe(name, expr, exprs: _*)
+  }
+
+  /**
+   * Waits for the first action on the observed dataset to complete and returns true.
+   * The result is then available through the get method.
+   * This method times out after the given amount of time returning false.
+   *
+   * @param time timeout
+   * @param unit timeout time unit
+   * @return true if action completed within timeout, false otherwise
+   * @throws InterruptedException interrupted while waiting
+   */
+  def waitCompleted(time: Long, unit: TimeUnit): Boolean = waitCompleted(Some(unit.toMillis(time)))
+
+  /**
+   * Get the observed metrics. This waits until the observed dataset finishes its first action.
+   * If you want to wait for the result and provide a timeout, use [[waitCompleted]]. Only the
+   * result of the first action is available. Subsequent actions do not modify the result.
+   *
+   * @return the observed metrics as a [[Row]]
+   * @throws InterruptedException interrupted while waiting
+   */
+  def get: Row = {
+    assert(waitCompleted(None), "waitCompleted without timeout returned false")
+    row.get
+  }
+
+  private def waitCompleted(millis: Option[Long]): Boolean = {
+    synchronized {
+      if (row.isEmpty) {
```

Comment:
You are saying `wait` can return before the given time is up and without `notify` being called? Or do you refer to `notify` being called when `row` being `row.isEmpty`? We could guarantee that this is not happening if you would use a synchronization object other than `this`.

Reply:
I have moved the `synchronized` to a private object to have full control over who is calling `notifyAll` on it. Now it is only called when `row.isDefined`, so that `wait()` only ever returns when `row.isDefined`.